python/dolfinx/wrappers/geometry.cpp

```diff
@@ -12,274 +12,185 @@
-        }
-        else if (gdim == 1 and px.ndim() == 1)
-        {
-          // 1D problem
-          for (py::ssize_t i = 0; i < px.shape(0); i++)
-            _p[3 * i] = px(i);
-        }
-        else if (px.ndim() == 2)
-        {
-          for (py::ssize_t i = 0; i < px.shape(0); i++)
-            for (py::ssize_t j = 0; j < px.shape(1); j++)
-              _p[3 * i + j] = px(i, j);
-        }
-        else
-          throw std::runtime_error("Array has wrong ndim.");
-
         return dolfinx::geometry::compute_colliding_cells<T>(
-            mesh, candidate_cells, _p);
+            mesh, candidate_cells, std::span(points.data(), points.size()));
       },
-      py::arg("mesh"), py::arg("candidate_cells"), py::arg("points"));
+      nb::arg("mesh"), nb::arg("candidate_cells"), nb::arg("points"));
 
   m.def(
       "compute_distance_gjk",
-      [](const py::array_t<T>& p, const py::array_t<T>& q)
+      [](nb::ndarray<const T, nb::c_contig> p,
+         nb::ndarray<const T, nb::c_contig> q)
       {
-        const std::size_t p_s0 = p.ndim() == 1 ? 1 : p.shape(0);
-        const std::size_t q_s0 = q.ndim() == 1 ? 1 : q.shape(0);
-        std::vector<T> _p(3 * p_s0), _q(3 * q_s0);
-
-        auto px = p.unchecked();
-        if (px.ndim() == 1)
-        {
-          for (py::ssize_t i = 0; i < px.shape(0); i++)
-            _p[i] = px(i);
-        }
-        else if (px.ndim() == 2)
-        {
-          for (py::ssize_t i = 0; i < px.shape(0); i++)
-            for (py::ssize_t j = 0; j < px.shape(1); j++)
-              _p[3 * i + j] = px(i, j);
-        }
-        else
-          throw std::runtime_error("Array has wrong ndim.");
-
-        auto qx = q.unchecked();
-        if (qx.ndim() == 1)
-        {
-          for (py::ssize_t i = 0; i < qx.shape(0); i++)
-            _q[i] = qx(i);
-        }
-        else if (qx.ndim() == 2)
-        {
-          for (py::ssize_t i = 0; i < qx.shape(0); i++)
-            for (py::ssize_t j = 0; j < qx.shape(1); j++)
-              _q[3 * i + j] = qx(i, j);
-        }
-        else
-          throw std::runtime_error("Array has wrong ndim.");
-
-        const std::array<T, 3> d
-            = dolfinx::geometry::compute_distance_gjk<T>(_p, _q);
-        return py::array_t<T>(3, d.data());
+        std::size_t p_s0 = p.ndim() == 1 ? 1 : p.shape(0);
+        std::size_t q_s0 = q.ndim() == 1 ? 1 : q.shape(0);
+        std::span<const T> _p(p.data(), 3 * p_s0), _q(q.data(), 3 * q_s0);
+        std::array<T, 3> d = dolfinx::geometry::compute_distance_gjk<T>(_p, _q);
+        return dolfinx_wrappers::as_nbarray_copy(d, {d.size()});
       },
-      py::arg("p"), py::arg("q"));
+      //   nb::rv_policy::copy,
+      nb::arg("p"), nb::arg("q"));
 
   m.def(
       "squared_distance",
       [](const dolfinx::mesh::Mesh<T>& mesh, int dim,
-         std::vector<std::int32_t> indices, const py::array_t<T>& points)
+         const std::vector<std::int32_t>& indices,
```

Comment:
Shoult this be rather nb::ndarray (to avoid the python list inefficiency)?

Reply:
Yes, and in quite a few mored places in the geometry interface. It wasn't done right before with pybind11, so this PR doesn't really change that.

We should check in nanobind is faster for `std::vector` than pybind11.

Reply:
I run a small benchmark. Nanobind is slightly (10%) faster for small arrays (<10^5), but for larger arrays this is simply the cost of `list(numpy_array)` that dominates, so no changes.